### ISSUE TITLE
(Re)run `terragrunt run-all destroy` even if some resources tore down successfully while others failed

### ISSUE THREAD

**Issue by karlschriek** [gruntwork-io/terragrunt#3183]
## Describe the bug

I am currently evaluating terragrunt's run-all capabilities and am getting unstuck on the `terragrunt run-all destroy`, in particular in the case of failure. I am not able to simply rerun this command if some modules destroyed correctly, but others failed. Let's say for example that `terragrunt run-all apply` generates the following dependency graph:

    Group 1
    - Module /root/myproject/base
    
    Group 2
    - Module /root/myproject/module1
    - Module /root/myproject/module2
    
    Group 3
    - Module /root/myproject/module1_extension

Above both "module1" and "module2" depend on "base"; "module1_extension" depends on "module1".

In this case `terragrunt run-all destroy` will destroy all the "Group 3" modules, then the "Group 2" modules, then the "Group 1" modules. In the case where it successfully tears down both "Group 3" and "Group 2", but fails to fully remove "Group 1" I was hoping to be able to just again run `terragrunt run-all destroy`, but this does not work, erroring out with:

    ERRO[0004] Module /root/myproject/module1_extension has finished with an error: /root/myproject/module1/terragrunt.hcl is a dependency of /root/myproject/module1_extension/terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block.  prefix=[/root/myproject/module1_extension]

## Steps To Reproduce

Steps to reproduce the behavior, code snippets and examples which can be used to reproduce the issue.

Create a project structure as above and then destroy  module1, module2 and module1_extension. If you now try to run "terragrunt run-all destroy" you will get the above error. (You can also just apply "module1" to begin with and then "run-all destroy")

## Expected behavior

To be clear, in this specific example only the "base" project still needs to be torn down (everything else is gone).  I could tear it down now with `--terragrunt-include-dir *base`, but since errors in destroy steps are quite common I would expect terragrunt to be able to deal with repeating the command to destroy all. Since destroy happens sequentially (in a reverse order to apply), terragrunt should know which states have been torn down and which haven't and should always be able to retry the ones that are still there (since, unless you specified `--terragrunt-ignore-dependency-errors`, it would not have gone on and torn down modules that anything that hasn't been torn down yet depends on).

EDIT: I am not entirely sure about the "unless you specified `--terragrunt-ignore-dependency-errors`" part. Would terragrunt go ahead and destroy "module1" even if "module1_extension" failed to destroy correctly with `--terragrunt-ignore-dependency-errors` set? Because if it still respects that dependency then just setting that flag is probably the solution to this post (although somewhat messy and full of irrelevant warnings).

## Nice to haves

- [ ] Terminal output
- [ ] Screenshots

## Versions

- Terragrunt version: 0.58.10
- OpenTofu/Terraform version: tf 1.8.3
- Environment details (Ubuntu 20.04, Windows 10, etc.):

**Comment by denis256** [comment]
Hello,
Terragrunt is simply the runner of Terraform and doesn't have transactional capabilities to execute all or nothing, once dependency was destroyed, outputs reading will not work.

It can be also used `--terragrunt-strict-include` and specify path to specific depdencnies to use them in destroy
https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-strict-include

**Comment by karlschriek** [comment]
But there are no dependencies that are needed for the destroy that doesn't
work

Let's say Module B depends on Module A.

I cannot apply Module B before applying module A. I also cannot destory
Module A before destroying Module B. That makes sense.

What doesn't make sense is that I cannot (using the run-all directive)
destroy Module A after I already destroyed Module B.

**Comment by denis256** [comment]
Hello,
in the current implementation, Terrgrunt "doesn't know" what was already destroyed and if the module should be skipped or not, so it is trying to do destroy each time.

**Comment by karlschriek** [comment]
Yes, that is clear. I am pointing out that that is very problematic in a
non-interactive setting where you want to be able to robust tear down
everything, with automatic retries etc. As it is right now if even just one
module tears down correctly and the next one fails you will not be able to
rerun the destroy

**Comment by yhakbar** [comment]
Hey @karlschriek,

This issue was closed out as stale, but I'm not sure it received satisfactory resolution. I also don't fully grasp the problem here.

Do you mind restating what you hoped Terragrunt would do, and why existing tooling from [runtime controls](https://terragrunt.gruntwork.io/docs/features/runtime-control/) are not sufficient to handle this edge case?

**Comment by fredleger** [comment]
please reopen

**Comment by yhakbar** [comment]
@fredleger can you supply those details?

**Comment by fredleger** [comment]
Same as others stated.
I wanted to destroy a full infrastructure created with terragrunt run-all apply. One component was destroyed outside of terraform and thus I was obliged to destroy everything by hand.

**Comment by yhakbar** [comment]
You're not, though. Terragrunt has tooling to make it so that you can mock dependencies that are missing, or retry errors during destroys, etc. so I'm not understanding what the gap is in Terragrunt functionality.

**Comment by fredleger** [comment]
ok can you details the procedure @yhakbar ? i did not find  it (like the others i guess) looking in terragrunt documentation

**Comment by yhakbar** [comment]
There really can't be because of how OpenTofu/Terraform work right now.

If, for example, you want to mock an AWS ARN that's acquired as an output from one unit, and used as an input to another unit, the AWS provider might require that the ARN be well formed (e.g. `arn:aws:s3:::my-s3-bucket`). Terragrunt can't just generically mock that value (using something like `my-mock-value`) without the user supplying what they know is a valid value, as using an arbitrary value can cause errors for the OpenTofu/Terraform AWS provider if it's malformed.

Terragrunt can't know how you plan on using these inputs or outputs in OpenTofu/Terraform, so you're required to explicitly configure mocks to prevent errors being thrown by providers and modules.

There's no way to tell OpenTofu/Terraform "Hey, I know you require this value, but I don't know what it is yet. Be OK with that for now, I promise it'll be available when it's needed, like for an apply." The OpenTofu team is interested in tackling this, but until then, users have to explicitly mock unknown outputs.

**Comment by nford** [comment]
The context of this ticket is specifically for destroying all resources though. There are no valid ARNs to provide. So I can't delete an IAM policy referencing a resource that has already deleted because the deleted resource needs to exist to provide a valid ARN to the one I'm deleting. I completely understand the logic in the context of apply, not destroy..

**Comment by tux3** [comment]
So, I hit this problem today, and I'd like to propose an idea for a different solution.

Although there are possible things like mock_outputs, it does not work in every situation, it's easy to misuse, and it is a lot of manual work to invest in a workaround.

Today when Terragrunt destroys multiple groups of units that have internal dependencies on each other, if the destroy is interrupted, as a user I am currently left in a terrible situation, because I have to finish the destroy by hand, unit by unit.  

---

When resuming an interrupted destroy, as a user I look back at what Terragrunt did to see what was and wasn't destroyed. Then I rebuild the DAG of *only* not-fully-destroyed units. But the point of that is *not* to skip unnecessary work, the idea is that Terragrunt can only start to destroy a unit's reverse dependencies *after this unit is fully destroyed*. This means skipping fully destroyed units is enough to resume an interrupted destroy correctly. If we can do this automatically, the user won't be left to hold both pieces.

The immediate objection is that Terragrunt *doesn't know* what state a unit is in without running a plan. So the proposal is to allow resuming a destroy by saving just enough when interrupted to keep track of which units were fully destroyed. 

There are different places where Terragrunt could save this flag. You could imagine a flag per unit, cleared by an attempt to start an apply. Or you could decide to keep it as a list in a temporary resume file that doesn't interact with other parts of Terragrunt, to keep the fix simple and localized.


I should also note that there is on paper a possible race, so we could want to ask the user to confirm before resuming a destroy. If we're resuming from where we left off, by construction we are skipping some plans, so someone could have re-applied those units that we were in the middle of destroying. They would have had to accept a scary looking plan, but it's possible. 

However the same race exists between a `terragrunt plan --all` and a `terragrunt apply --all`. There's a confirm prompt there too, so it's nothing unprecedented.

As a general piece of feedback, I really like Terragrunt, but I cannot convey how unfortunate it feels that there is this asymmetry between applying and destroying. Everything naturally works well when applying, but Terragrunt could also make destroy operations much friendlier by keeping better track of its own state (stack, units), in the same way that Terraform keep track of Terraform state (resources).

**Comment by yhakbar** [comment]
@tux3 

I understand your frustration. I do actually think there would be a way to introduce the kind of resumption that you're asking for here, as Terragrunt has the experimental ability to record a ledger of what happened in a run via the [Run Report](https://terragrunt.gruntwork.io/docs/features/run-report/).

We could, hypothetically, do something with that to allow resumption on only failed units, but we would still have the problem that the dependent unit no longer has OpenTofu/Terraform state, and thus can't provide outputs.

**Comment by karlschriek** [comment]
Just to come full circle on this. When I initially opened this issue I was busy evaluating whether Terragrunt could help me with various problems I was pushing up against when managing infrastructure with Terraform, the biggest one being that I needed a way to break deployments down into smaller modules while still maintaining dependencies. On the surface of it, that is exactly what Terragrunt does with the dependency blocks and `run-all ` feature, but since it all executes as a single shell session the only context that is available is whatever is stored in the single codebase you have locally. It became clear to me pretty quickly that you would need an approach the works with a remote orchestrator if you want to do this properly.

So I went off and built Snap CD. You can find its public GitHub repo here: https://github.com/schrieksoft/snapcd
